### PR TITLE
Implement resetOption method

### PR DESCRIPTION
**Pull request by dxvladislavvolkov**

### CODE REVIEW

File: testing/tests/DevExpress.core/component.tests.js [hunk 1]
```diff
@@ -1302,6 +1311,40 @@ QUnit.module("defaultOptions", {
...
+        assert.equal(instance.option("opt5.subOpt1"), "someValue");
+        assert.equal(instance.option("opt5.subOpt2"), "someValue");
+
+        instance.resetOption('opt4');
+        instance.resetOption('opt5');
+
+        assert.equal(instance.option("opt4"), "default");
+        assert.equal(instance.option("opt5.subOpt1"), "default");
+        assert.equal(instance.option("opt5.subOpt2"), undefined);
+    });
```
Comment (nightskylark) [inline testing/tests/DevExpress.core/component.tests.js:1358]: Does the `resetOption` method support resetting the second-level options like `instance.resetOption('opt5.subopt1')`?

File: js/core/utils/option.js [hunk 2]
```diff
@@ -0,0 +1,145 @@
+import coreDataUtils from "./data";
```
Comment (nightskylark) [inline js/core/utils/option.js]: We should not move any extracted modules to the `utils` folder, The `utils` means that this methods do not consider any context. It should be only simple methods with straight in-out logic and withot any coherence. I think, it's better to rename this module and place it to the `core` folder.

File: js/core/utils/option.js [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+        functionsAsIs: true,
+        merge,
+        unwrapObservables: path.length > 1 && !!this._getOptionsByReference()[path[0]]
+    });
+};
+
+const setOption = function(name, value, merge) {
+    const previousValue = getOptionValue.bind(this)(this._options, name, false);
+
+    if(this._optionValuesEqual(name, previousValue, value)) {
```
Comment (nightskylark) [inline js/core/utils/option.js]: Should this method be always called with the widget instance as a context? If so, It definitely needs to be refactored.

File: js/core/dom_component.js [hunk 4]
```diff
@@ -172,8 +172,8 @@ var DOMComponent = Component.inherit({
     _renderDimensions: function() {
         var $element = this.$element();
         var element = $element.get(0);
-        var width = this._getOptionValue("width", element);
-        var height = this._getOptionValue("height", element);
+        var width = this._computeOptionValue("width", element);
```
Comment (nightskylark) [inline js/core/dom_component.js]: Why? I did not find any conflicts.

File: js/core/utils/option.js (hunk 2 above)
Comment (nightskylark) [reply js/core/utils/option.js]: It's great that you extract the code but it looks like the work is not finished yet. Let's discuss it personally.

**Review by nightskylark — CHANGES_REQUESTED** [review]

File: js/core/option_manager.js [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+import typeUtils from "./utils/type";
+import domAdapter from "./dom_adapter";
+
+export class OptionManager {
+    constructor(options, optionsByReference, logDeprecatedWarning, deprecatedOptions) {
+        this._options = options;
+        this._optionsByReference = optionsByReference;
+        this._deprecatedOptions = deprecatedOptions;
+        this._logDeprecatedWarning = logDeprecatedWarning;
+        this._deprecatedOptionsSuppressed;
```
Comment (nightskylark) [inline js/core/option_manager.js]:
```suggestion
        this._deprecatedOptionsSuppressed = false;
```

File: js/core/component.js [hunk 6]
```diff
@@ -264,6 +264,14 @@ var Component = Class.inherit({
         }.bind(this));
     },
 
+    _logDeprecatedWarningCount: 0,
+
+    _logDeprecatedWarning(option, info) {
+        var message = info.message || ("Use the '" + info.alias + "' option instead");
+        errors.log("W0001", this.NAME, option, info.since, message);
+        ++this._logDeprecatedWarningCount;
```
Comment (nightskylark) [inline js/core/component.js]: This code was written for testing only and needs to be removed.

File: js/core/component.js [hunk 7]
```diff
@@ -241,13 +241,12 @@ var Component = Class.inherit({
             this._optionManager.onChanged((name, value, previousValue) => {
                 this._notifyOptionChanged(name, value, previousValue);
             });
-            this._optionManager.suppressDeprecatedWarnings();
+
```
Comment (nightskylark) [inline js/core/component.js:237]: Redundant line

File: js/core/component.js [hunk 8]
```diff
@@ -205,6 +209,7 @@ var Component = Class.inherit({
             this.setEventsStrategy(options.eventsStrategy);
         }
         this._options = {};
+        this._optionManager = {};
```
Comment (nightskylark) [inline js/core/component.js]: What is it for?

File: js/core/option_manager.js [hunk 9]
```diff
@@ -0,0 +1,240 @@
...
+
+    onChanging(callBack) {
+        this._changingCallbacks.add(callBack);
+    }
+
+    onChanged(callBack) {
+        this._changedCallbacks.add(callBack);
+    }
+
+    onLogWarning(callBack) {
```
Comment (nightskylark) [inline js/core/option_manager.js]: It looks like this method and internal callbacks should be renamed. This module know nothing about logging warnings. It just noticed about the deprecated option was processed.

File: js/core/option_manager.js [hunk 10]
```diff
@@ -0,0 +1,240 @@
...
+        this._deprecatedOptions = deprecatedOptions;
+        this._changingCallbacks = CallBacks({ syncStrategy: true });
+        this._changedCallbacks = CallBacks({ syncStrategy: true });
+        this._logWarningCallbacks = CallBacks({ syncStrategy: true });
+        this._cachedDeprecateNames = [];
+        this.cachedGetters = {};
+        this.cachedSetters = {};
+    }
+
+    _valuesEqual(name, oldValue, newValue) {
```
Comment (nightskylark) [inline js/core/option_manager.js]: It looks like first argument `name` is not used

File: js/integration/angular/component_registrator.js [hunk 11]
```diff
@@ -163,7 +163,7 @@ let ComponentBuilder = Class.inherit({
                 this._component.option(optionPath, newValue);
                 updateWatcher();
 
-                if(this._component._optionValuesEqual(optionPath, oldValue, newValue) && this._ngLocker.locked(optionPath)) {
+                if(this._component._optionManager._valuesEqual(optionPath, oldValue, newValue) && this._ngLocker.locked(optionPath)) {
```
Comment (nightskylark) [inline js/integration/angular/component_registrator.js]: So much underscores. Maybe this method should be extracted as an utility?

File: js/core/component.js [hunk 12]
```diff
@@ -504,12 +477,22 @@ var Component = Class.inherit({
         return Object.prototype.hasOwnProperty.call(deprecatedOptions, name);
     },
 
-    _setOptionSilent: function(name, value) {
+    _setOptionWithoutNotify: function(name, value) {
```
Comment (nightskylark) [inline js/core/component.js]: I suppose this renaming and adding another method wtih the old name could confuse our collegues. Maybe it's better to find another names for the new methods or use the `silent` flag as argument?

**Review by nightskylark — CHANGES_REQUESTED** [review]

File: testing/tests/DevExpress.core/component.tests.js (hunk 1 above)
Comment (ovchinnikov) [reply testing/tests/DevExpress.core/component.tests.js:1358]: I suggest to check "items[1].subOption" case as well

File: js/core/component.js [hunk 13]
```diff
@@ -383,21 +351,21 @@ var Component = Class.inherit({
...
-            currentOptions = this._options;
-            this._options = {};
-            this._initialized = false;
-            this._setDefaultOptions();
-            this._setOptionsByDevice(currentOptions.defaultOptionsRules);
-
-            this._initialOptions = this._options;
-            this._options = currentOptions;
-            this._initialized = currentInitialized;
+            this._initialOptions = {};
```
Comment (nightskylark) [inline js/core/component.js]: Needs to be removed. Empty object will be overwritten in next line.

File: js/core/component.js [hunk 14]
```diff
@@ -383,21 +351,21 @@ var Component = Class.inherit({
...
+            this._initialOptions = this._getDefaultOptions();
+            const rulesOptions = this._getOptionByRules(this._getOptionSilent("defaultOptionsRules"));
```
Comment (nightskylark) [inline js/core/component.js]: Why do you use `this._getOptionSilent` and `_optionManager._setSilent` at the same time. It looks like component's methods should be used everywhere

File: js/core/component.js [hunk 15]
```diff
@@ -490,6 +456,14 @@ var Component = Class.inherit({
         return result;
     },
 
+    _getOptionSilent: function(name) {
```
Comment (nightskylark) [inline js/core/component.js]: Maybe it should be `_getOptionByStealth`? I think, pairy methods should be named the same way.

File: js/core/option_manager.js [hunk 16]
```diff
@@ -0,0 +1,215 @@
...
+        this._deprecatedOptions = deprecatedOptions;
+        this._changingCallbacks = CallBacks({ syncStrategy: true });
+        this._changedCallbacks = CallBacks({ syncStrategy: true });
+        this._deprecatedCallbacks = CallBacks({ syncStrategy: true });
+        this._cachedDeprecateNames = [];
+        this.cachedGetters = {};
+        this.cachedSetters = {};
+    }
+
+    _logWarningIfDeprecated(option) {
```
Comment (nightskylark) [inline js/core/option_manager.js]: Why you renamed the callback but left the `log...` name for the method? I suggest something like `notifyDeprecated`.

File: js/core/utils/compare.js [hunk 17]
```diff
@@ -0,0 +1,30 @@
+import domAdapter from "../dom_adapter";
+import coreDataUtils from "./data";
+import typeUtils from "./type";
+
+const valuesEqual = function(oldValue, newValue) {
```
Comment (nightskylark) [inline js/core/utils/compare.js]: I suggest to move this method to `data` utility file. The `compare` filename looks too mysterious without a context.

File: js/core/utils/compare.js [hunk 18]
```diff
@@ -0,0 +1,30 @@
...
+    }
+
+    const oldValueIsNaN = oldValue !== oldValue;
+    const newValueIsNaN = newValue !== newValue;
+    if(oldValueIsNaN && newValueIsNaN) {
+        return true;
+    }
+
+    if(oldValue === 0 && newValue === 0) {
+        return (1 / oldValue) === (1 / newValue);
```
Comment (nightskylark) [inline js/core/utils/compare.js]: Maybe we need to extract some variable or method to explain negative zero situation. I think this code could rise a questions for anyone who will see it.

File: js/core/component.js [hunk 19]
```diff
@@ -383,21 +351,21 @@ var Component = Class.inherit({
...
-            this._setOptionsByDevice(currentOptions.defaultOptionsRules);
-
-            this._initialOptions = this._options;
-            this._options = currentOptions;
-            this._initialized = currentInitialized;
+            this._initialOptions = {};
+            this._initialOptions = this._getDefaultOptions();
+            const rulesOptions = this._getOptionByRules(this._getOptionSilent("defaultOptionsRules"));
+            this._optionManager._setSilent(this._initialOptions, rulesOptions);
+        }
```
Comment (nightskylark) [inline js/core/component.js:367]: Why don't we call `_setOptionsByDevice` here?

File: js/core/component.js [hunk 20]
```diff
@@ -532,173 +516,28 @@ var Component = Class.inherit({
...
-            return value.bind(context)();
+        try {
+            this._optionManager.setValue(options, value);
+        } finally {
+            this.endUpdate();
         }
+    },
 
-        return value;
+    resetOption: function(name) {
```
Comment (nightskylark) [inline js/core/component.js:552]: The doctags for `resetOption` method should be added here.

**Review by nightskylark — CHANGES_REQUESTED** [review]

File: js/core/option_manager.js [hunk 21]
```diff
@@ -0,0 +1,215 @@
+import coreDataUtils from "./utils/data";
+import typeUtils from "./utils/type";
+import compareUtils from "./utils/compare";
+import CallBacks from "./utils/callbacks";
```
Comment (IlyaKhD) [inline js/core/option_manager.js]:
```suggestion
import createCallBack from "./utils/callbacks";
```

File: js/core/option_manager.js [hunk 22]
```diff
@@ -0,0 +1,215 @@
...
+        options = this._ensureObject(options, value);
+
+        this._setSilent(this._options, options);
+    }
+
+    getSilent(name) {
+        return this._options[name];
+    }
+
+    setValue(options, value) {
```
Comment (IlyaKhD) [inline js/core/option_manager.js]: Consider reordering:
`getValue`
`setValue`
`getSilent`
`setSilent`

File: js/core/option_manager.js [hunk 23]
```diff
@@ -0,0 +1,215 @@
...
+    onDeprecated(callBack) {
+        this._deprecatedCallbacks.add(callBack);
+    }
+
+    getValue(name) {
+        const normalizedName = this._normalizeName(name);
+        return this._getValue(this._options, normalizedName);
+    }
+
+    setSilent(options, value) {
```
Comment (IlyaKhD) [inline js/core/option_manager.js]:
```suggestion
    setValueSilently(options, value) {
```

File: js/core/option_manager.js [hunk 24]
```diff
@@ -0,0 +1,215 @@
...
+        options = this._ensureObject(options, value);
+
+        this._setSilent(this._options, options);
+    }
+
+    getSilent(name) {
```
Comment (IlyaKhD) [inline js/core/option_manager.js]:
```suggestion
    getValueSilently(name) {
```

File: js/core/option_manager.js [hunk 25]
```diff
@@ -0,0 +1,215 @@
...
+        this._changedCallbacks.add(callBack);
+    }
+
+    onDeprecated(callBack) {
+        this._deprecatedCallbacks.add(callBack);
+    }
+
+    getValue(name) {
+        const normalizedName = this._normalizeName(name);
+        return this._getValue(this._options, normalizedName);
```
Comment (IlyaKhD) [inline js/core/option_manager.js]: No need to extract a variable
```suggestion
        return this._getValue(this._options, this._normalizeName(name));
```

File: js/core/option_manager.js [hunk 26]
```diff
@@ -0,0 +1,215 @@
...
+            for(const valueName in value) {
+                this._prepareRelevantNames(options, name + "." + valueName, value[valueName]);
+            }
+        }
+
+        this._setRelevantNames(options, name, value);
+    }
+
+    _ensureObject(options, value) {
+        let name = options;
```
Comment (IlyaKhD) [inline js/core/option_manager.js]: I propose avoiding argument reassigning if possible.

```
_convertToObj(option, value) {
    if(typeof option !== "string")
        return option;

    const result = {};
    result[name] = value;
    return result;
}
````

File: js/core/option_manager.js [hunk 27]
```diff
@@ -0,0 +1,215 @@
...
+
+        this._setSilent(this._options, options);
+    }
+
+    getSilent(name) {
+        return this._options[name];
+    }
+
+    setValue(options, value) {
+        options = this._ensureObject(options, value);
```
Comment (IlyaKhD) [inline js/core/option_manager.js]: As I got it - the `options` argument can be either a `string` or a an `object`. Is this really required? Can't we use a stricter signature  (e.g. accepting `object`-s only)

File: js/core/utils/compare.js [hunk 28]
```diff
@@ -0,0 +1,30 @@
...
+    }
+
+    if(oldValue === null || typeof oldValue !== "object" || domAdapter.isElementNode(oldValue)) {
+        return oldValue === newValue;
+    }
+
+    return false;
+};
+
+exports.valuesEqual = valuesEqual;
```
Comment (IlyaKhD) [inline js/core/utils/compare.js]:
```suggestion
exports.equals= equals;
```

File: js/core/option_manager.js [hunk 29]
```diff
@@ -0,0 +1,215 @@
...
+
+    _clearField(options, name) {
+        delete options[name];
+
+        const previousFieldName = this._getParentName(name);
+        const fieldName = this._getFieldName(name);
+        const fieldObject = previousFieldName ? this._getValue(options, previousFieldName, false) : options;
+
+        if(fieldObject) {
+            delete fieldObject[fieldName];
```
Comment (IlyaKhD) [inline js/core/option_manager.js]: No need to have a variable
```suggestion
            delete fieldObject[this._getFieldName(name)];
```

File: js/core/option_manager.js [hunk 30]
```diff
@@ -0,0 +1,215 @@
...
+        }
+    }
+
+    _getParentName(fullName) {
+        const splitNames = fullName.split('.');
```
Comment (IlyaKhD) [inline js/core/option_manager.js]:
```suggestion
        return fullName.substr(0, fullName.lastIndexOf('.'));
```

File: js/core/option_manager.js [hunk 31]
```diff
@@ -0,0 +1,215 @@
...
+        }
+
+        this._changingCallbacks.fire(name, previousValue, value);
+
+        this._setValue(name, value, merge);
+        this._changedCallbacks.fire(name, value, previousValue);
+    }
+
+    _setRelevantNames(options, name, value) {
+        if(name) {
```
Comment (IlyaKhD) [inline js/core/option_manager.js]:
```suggestion
        if(!name) return;
```

File: js/core/option_manager.js [hunk 32]
```diff
@@ -0,0 +1,215 @@
...
+
+            if(alias && alias !== name) {
+                this._setField(options, alias, value);
+                this._clearField(options, name);
+            }
+        }
+    }
+
+    _normalizeName(name) {
+        if(name) {
```
Comment (IlyaKhD) [inline js/core/option_manager.js]:
```suggestion
        if(!name) return name;
```

**Review by IlyaKhD — CHANGES_REQUESTED** [review]

**Review by nightskylark — DISMISSED** [review]

File: js/core/component.js [hunk 33]
```diff
@@ -383,21 +352,20 @@ var Component = Class.inherit({
...
+            this._initialOptions = this._getDefaultOptions();
+            const rulesOptions = this._getOptionByRules(this._getOptionByStealth("defaultOptionsRules"));
+            this._optionManager.setValueByReference(this._initialOptions, rulesOptions);
         }
 
-        return this._initialOptions[optionName];
+        optionName = optionName.replace(/\[/g, ".").replace(/\]/g, "");
+        const fullPath = optionName.split(".");
+        let value;
+        for(let path of fullPath) {
```
Comment (IlyaKhD) [inline js/core/component.js]:
```suggestion
        for(const path of optionName.split(".")) {
```

File: js/core/component.js [hunk 34]
```diff
@@ -490,6 +456,23 @@ var Component = Class.inherit({
...
+        return result;
+    },
+
+    _getOptionByStealth: function(name) {
+        return this._optionManager.getValueSilently(name);
+    },
+
+    _setOptionByStealth: function(options, value) {
+        options = this._convertToObj(options, value);
+        this._optionManager.setValueSilently(options);
```
Comment (IlyaKhD) [inline js/core/component.js]: No need to reassign the attribute:
```suggestion
        this._optionManager.setValueSilently(this._convertToObj(options, value));
```

File: js/core/component.js [hunk 35]
```diff
@@ -490,6 +456,23 @@ var Component = Class.inherit({
         return result;
     },
 
+    _convertToObj(options, value) {
```
Comment (IlyaKhD) [inline js/core/component.js]: Better make this a static function outside the class (no need to use `this` context)

File: js/core/component.js [hunk 36]
```diff
@@ -532,173 +525,39 @@ var Component = Class.inherit({
...
+     * @param1 optionName:string
+     */
+    resetOption: function(name) {
+        if(!name) {
+            return;
+        }
+        const defaultValue = this.initialOption(name);
+        this.beginUpdate();
+        const option = this._convertToObj(name, defaultValue);
+        this._optionManager.setValue(option);
```
Comment (IlyaKhD) [inline js/core/component.js]: No need to use a variable

```suggestion
        this._optionManager.setValue(this._convertToObj(name, defaultValue));
```

File: js/core/component.js (hunk 35 above)
Comment (IlyaKhD) [reply js/core/component.js]: What about renaming it to smth like `normalizeOptions` or `normalizeOptionsObj` ?

File: js/core/option_manager.js [hunk 37]
```diff
@@ -0,0 +1,195 @@
...
+            delete fieldObject[this._getFieldName(name)];
+        }
+    }
+
+    _getParentName(fullName) {
+        return fullName.substr(0, fullName.lastIndexOf('.'));
+    }
+
+    _getFieldName(fullName) {
+        const splitNames = fullName.split('.');
```
Comment (IlyaKhD) [inline js/core/option_manager.js]:
```suggestion
         return fullName.substr(fullName.lastIndexOf('.') + 1);
```

File: js/core/option_manager.js [hunk 38]
```diff
@@ -0,0 +1,195 @@
...
+
+        this._changingCallbacks.fire(name, previousValue, value);
+
+        this._setValue(name, value, merge);
+        this._changedCallbacks.fire(name, value, previousValue);
+    }
+
+    _setRelevantNames(options, name, value) {
+        if(!name) return;
+        const alias = this._normalizeName(name);
```
Comment (IlyaKhD) [inline js/core/option_manager.js]: As for me `alias` has a different meaning from `normalize`. I propose to use only one word i.e. rename variable to `normalizedName` or function to `getAlias`

File: js/core/option_manager.js [hunk 39]
```diff
@@ -0,0 +1,195 @@
...
+                this._cachedDeprecateNames.push(optionName);
+            }
+        }
+        for(let i = 0; i < this._cachedDeprecateNames.length; i++) {
+            if(this._cachedDeprecateNames[i] === name) {
+                deprecate = this._deprecatedOptions[name];
+                break;
+            }
+        }
+        if(deprecate) {
```
Comment (IlyaKhD) [inline js/core/option_manager.js:115]: No need to have a variable

```diff
-	        const alias = deprecate.alias;

-            if(alias ) {
+            if(deprecate.alias) {
-                name = alias;
+                name = deprecate.alias;
            }
```

File: js/core/utils/data.js [hunk 40]
```diff
@@ -200,6 +201,92 @@ var toComparable = function(value, caseSensitive) {
     return value;
 };
 
+const isArraysEqualByValue = function(array1, array2, deep) {
```
Comment (IlyaKhD) [inline js/core/utils/data.js]:
```suggestion
const arraysEqualByValue = function(array1, array2, depth) {
```

File: js/core/utils/data.js [hunk 41]
```diff
@@ -200,6 +201,92 @@ var toComparable = function(value, caseSensitive) {
...
+    for(let i = 0; i < array1.length; i++) {
+        if(!equalByComplexValue(array1[i], array2[i], deep + 1)) {
+            return false;
+        }
+    }
+
+    return true;
+};
+
+const isObjectsEqualByValue = function(object1, object2, deep) {
```
Comment (IlyaKhD) [inline js/core/utils/data.js]:
```suggestion
const objectsEqualByValue = function(object1, object2, depth) {
```

File: js/core/utils/data.js [hunk 42]
```diff
@@ -200,6 +201,92 @@ var toComparable = function(value, caseSensitive) {
...
+    for(const propertyName in object2) {
+        if(!(propertyName in object1)) {
+            return false;
+        }
+    }
+
+    return true;
+};
+
+const maxEqualityDeep = 3;
```
Comment (IlyaKhD) [inline js/core/utils/data.js]:
```suggestion
const maxEqualityDepth = 3;
```

**Review by IlyaKhD — CHANGES_REQUESTED** [review]

**Review by IlyaKhD — DISMISSED** [review]

**Review by ovchinnikov — DISMISSED** [review]

**Review by nightskylark — APPROVED** [review]

**Pull request merged**